clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -2202,40 +2198,51 @@ static void printTemplateArguments(ASTContext &Ctx, raw_ostream &ArgOS,
   ArgOS << ">";
 }
 
-static std::string getKernelNameTypeString(QualType T, ASTContext &Ctx,
-                                           const PrintingPolicy &TypePolicy) {
-
-  QualType FullyQualifiedType = TypeName::getFullyQualifiedType(T, Ctx, true);
-
-  const CXXRecordDecl *RD = T->getAsCXXRecordDecl();
-
-  if (!RD)
-    return eraseAnonNamespace(FullyQualifiedType.getAsString(TypePolicy));
-
-  // If kernel name type is a template specialization with enum type
-  // template parameters, enumerators in name type string should be
-  // replaced  with their underlying value since the enum definition
-  // is not visible in integration header.
-  if (const auto *TSD = dyn_cast<ClassTemplateSpecializationDecl>(RD)) {
-    SmallString<64> Buf;
-    llvm::raw_svector_ostream ArgOS(Buf);
-
-    // Print the qualifiers for the type.
-    FullyQualifiedType.getQualifiers().print(ArgOS, TypePolicy,
+static std::string printRecordType(QualType T, const CXXRecordDecl *RD,
+                                   const PrintingPolicy &TypePolicy) {
+  SmallString<64> Buf;
+  llvm::raw_svector_ostream OS(Buf);
+  T.getCanonicalType().getQualifiers().print(OS, TypePolicy,
                                              /*appendSpaceIfNotEmpty*/ true);
+  if (const auto *TSD = dyn_cast<ClassTemplateSpecializationDecl>(RD)) {
 
     // Print template class name
-    TSD->printQualifiedName(ArgOS, TypePolicy, /*WithGlobalNsPrefix*/ true);
+    TSD->printQualifiedName(OS, TypePolicy, /*WithGlobalNsPrefix*/ true);
 
     // Print template arguments substituting enumerators
     ASTContext &Ctx = RD->getASTContext();
     const TemplateArgumentList &Args = TSD->getTemplateArgs();
-    printTemplateArguments(Ctx, ArgOS, Args.asArray(), TypePolicy);
+    printTemplateArguments(Ctx, OS, Args.asArray(), TypePolicy);
+
+    return eraseAnonNamespace(OS.str().str());
+  }
+  if (const auto *D = dyn_cast<CXXRecordDecl>(RD)) {
+    if (RD->getDeclContext()->isFunctionOrMethod())
+      return eraseAnonNamespace(T.getCanonicalType().getAsString(TypePolicy));
+    if (RD->getDeclContext()->isNamespace()) {
+      auto *NS = dyn_cast<NamespaceDecl>(RD->getDeclContext());
+      if (NS->isAnonymousNamespace())
+        RD->printQualifiedName(OS, TypePolicy, false);
+      else
+        RD->printQualifiedName(OS, TypePolicy, true);
+    } else
+      RD->printQualifiedName(OS, TypePolicy, true);
+    return eraseAnonNamespace(OS.str().str());
+  }
+  return eraseAnonNamespace(T.getCanonicalType().getAsString(TypePolicy));
```

Comment:
A few suggestions to this place:
- No need to cast `RD` to `CXXRecordDecl`, it seems to be already a `CXXRecordDecl`, right?
- We can check that declaration context is a namespace using `dyn_cast`
- Just a bit less of ifs
```suggestion
  if (RD->getDeclContext()->isFunctionOrMethod())
    return eraseAnonNamespace(T.getCanonicalType().getAsString(TypePolicy));
  if (auto *NS = cast<NamespaceDecl>(RD->getDeclContext()))
    RD->printQualifiedName(OS, TypePolicy, !NS->isAnonymousNamespace());
  else
    RD->printQualifiedName(OS, TypePolicy, /*WithGlobalNsPrefix=*/true);
  return eraseAnonNamespace(OS.str().str());
```

Reply:
This one still is not resolved.